components/movementsensor/adxl345/adxl345.go

```diff
@@ -0,0 +1,298 @@
+		) (interface{}, error) {
+			return NewAdxl345(ctx, deps, cfg, logger)
+		},
+	})
+
+	config.RegisterComponentAttributeMapConverter(movementsensor.SubtypeName, modelName,
+		func(attributes config.AttributeMap) (interface{}, error) {
+			var attr AttrConfig
+			return config.TransformAttributeMapToStruct(&attr, attributes)
+		},
+		&AttrConfig{})
+}
+
+type adxl345 struct {
+	bus        board.I2C
+	i2cAddress byte
+	logger     golog.Logger
+
+	// Lock the mutex when you want to read or write either the acceleration or the last error.
+	mu                 sync.Mutex
+	linearAcceleration r3.Vector
+	lastError          error
+
+	// Used to shut down the background goroutine which polls the sensor.
+	backgroundContext       context.Context
+	cancelFunc              func()
+	activeBackgroundWorkers sync.WaitGroup
+
+	generic.Unimplemented // Implements DoCommand with an ErrUnimplemented response
+}
+
+// NewAdxl345 is a constructor to create a new object representing an ADXL345 accelerometer.
+func NewAdxl345(
+	ctx context.Context,
+	deps registry.Dependencies,
+	rawConfig config.Component,
+	logger golog.Logger,
+) (movementsensor.MovementSensor, error) {
+	cfg := rawConfig.ConvertedAttributes.(*AttrConfig)
+	b, err := board.FromDependencies(deps, cfg.BoardName)
+	if err != nil {
+		return nil, err
+	}
+	localB, ok := b.(board.LocalBoard)
+	if !ok {
+		return nil, errors.Errorf("board %s is not local", cfg.BoardName)
+	}
+	bus, ok := localB.I2CByName(cfg.I2cBus)
+	if !ok {
+		return nil, errors.Errorf("can't find I2C bus '%s' for ADXL345 sensor", cfg.I2cBus)
+	}
+
+	var address byte
+	if cfg.UseAlternateI2CAddress {
+		address = 0x1D
+	} else {
+		address = 0x53
+	}
+
+	backgroundContext, cancelFunc := context.WithCancel(ctx)
+
+	sensor := &adxl345{
+		bus:               bus,
+		i2cAddress:        address,
+		logger:            logger,
+		backgroundContext: backgroundContext,
+		cancelFunc:        cancelFunc,
+	}
+
+	// To check that we're able to talk to the chip, we should be able to read register 0 and get
+	// back the device ID (0xE5).
+	deviceID, err := sensor.readByte(ctx, 0)
+	if err != nil {
+		return nil, errors.Errorf("can't read from I2C address %d on bus %s of board %s: '%s'",
+			address, cfg.I2cBus, cfg.BoardName, err.Error())
+	}
+	if deviceID != 0xE5 {
+		return nil, errors.Errorf("unexpected I2C device instead of ADXL345 at address %d: deviceID '%d'",
+			address, deviceID)
+	}
+
+	// The chip starts out in standby mode. Set it to measurement mode so we can get data from it.
+	// To do this, we set the Power Control register (0x2D) to turn on the 8's bit.
+	err = sensor.writeByte(ctx, 0x2D, 0x08)
+	if err != nil {
+		return nil, errors.Errorf("unable to put ADXL345 into measurement mode: '%s'", err.Error())
+	}
+
+	// Now, turn on the background goroutine that constantly reads from the chip and stores data in
+	// the object we created.
+	sensor.activeBackgroundWorkers.Add(1)
+	utils.PanicCapturingGo(sensor.pollData)
```

Comment:
define the function here, we prefer to do it this way so no one else can call this function with later potential work on this driver

Reply:
Subtle but wise choice! I should also do that in the MPU code; should that be in this same PR or a separate PR?

Reply:
Nicolas and I discussed: it can go in this same PR. I've made similar changes re: helper functions in utils/math.go and reordering the select cases for the MPU6050 code, too.